src/main/java/persistence/entity/impl/EntityManagerImpl.java

```diff
@@ -5,40 +5,115 @@
 import persistence.entity.EntityLoader;
 import persistence.entity.EntityManager;
 import persistence.entity.EntityPersister;
+import persistence.entity.PersistenceContext;
+import persistence.sql.EntityMetadata;
 
 public class EntityManagerImpl implements EntityManager {
     private final EntityPersister entityPersister;
 
     private final EntityLoader entityLoader;
 
+    private final PersistenceContext persistenceContext;
+
     public EntityManagerImpl(JdbcTemplate jdbcTemplate) {
-        this(new EntityPersisterImpl(jdbcTemplate), new EntityLoaderImpl(jdbcTemplate));
+        this(new EntityPersisterImpl(jdbcTemplate), new EntityLoaderImpl(jdbcTemplate),
+            new PersistenceContextImpl());
     }
 
-    public EntityManagerImpl(EntityPersister entityPersister, EntityLoader entityLoader) {
+    public EntityManagerImpl(EntityPersister entityPersister, EntityLoader entityLoader,
+        PersistenceContext persistenceContext) {
         this.entityPersister = entityPersister;
         this.entityLoader = entityLoader;
+        this.persistenceContext = persistenceContext;
     }
 
     @Override
-    public <T> T find(Class<T> entityClass, Long id) {
-        return entityLoader.select(entityClass, id);
+    public <T> T find(Class<T> entityClass, Object id) {
+        T cachedEntity = persistenceContext.getEntity(entityClass, id);
+
+        if (cachedEntity != null) {
+            return cachedEntity;
+        }
+
+        T entity = entityLoader.select(entityClass, id);
+
+        persistenceContext.addEntity(id, entity);
+        persistenceContext.getDatabaseSnapshot(id, entity);
+
+        return entity;
     }
 
     @Override
     public <T> List<T> findAll(Class<T> entityClass) {
-        return entityLoader.selectAll(entityClass);
+        List<T> entities = entityLoader.selectAll(entityClass);
+
+        entities.forEach(entity -> {
+            EntityMetadata entityMetadata = new EntityMetadata(entity.getClass());
+
+            Object id = entityMetadata.getIdFrom(entity);
+
+            persistenceContext.addEntity(id, entity);
+            persistenceContext.getDatabaseSnapshot(id, entity);
+        });
+
+        return entities;
     }
 
     @Override
-    public Object persist(Object entity) {
-        entityPersister.insert(entity);
+    public <T> T persist(T entity) {
+        if (!isNew(entity)) {
+            save(entity);
+        } else {
+            merge(entity);
+        }
```

Comment:
```suggestion
        if (isNew(entity)) {
            merge(entity);
        } else {
            save(entity);
        }
```

부정문은 한번더 생각을 해야하기 때문에 혼란스러울수  있어요~